src/drivers/mediatek/mt8186/interrupt.c

```diff
@@ -0,0 +1,227 @@
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#define PENDING_IRQ_INDEX_MAX 32
+
+/* fa00558c-d653-4851-a03a-b21f125a9524 */
+DECLARE_SOF_UUID("irq-mt8186", irq_mt8186_uuid, 0xfa00558c, 0xd653, 0x4851,
+		 0xa0, 0x3a, 0xb2, 0x1f, 0x12, 0x5a, 0x95, 0x24);
+
+DECLARE_TR_CTX(int_tr, SOF_UUID(irq_mt8186_uuid), LOG_LEVEL_INFO);
+
+static void mtk_irq_init(void)
+{
+	/* disable all ADSP IRQ */
+	io_reg_write(MTK_ADSP_IRQ_EN, 0);
+
+	 /* mask all IRQs between ADSP and other subsys */
+	io_reg_write(MTK_ADSP_IRQ_MASK, MTK_DSP_OUT_IRQ_MASK);
+}
+
+static void mtk_irq_mask(struct irq_desc *desc, uint32_t irq, unsigned int core)
+{
+	if (!desc) {
+		io_reg_update_bits(MTK_ADSP_IRQ_EN, BIT(irq), 0);
+	} else {
+		switch (desc->irq) {
+		case MTK_DSP_IRQ_MAILBOX:
+			io_reg_update_bits(MTK_ADSP_IRQ_EN, BIT(desc->irq), 0);
+			break;
+		default:
+			tr_err(&int_tr, "Invalid interrupt %d", desc->irq);
+			return;
+		}
+	}
+}
+
+static void mtk_irq_unmask(struct irq_desc *desc, uint32_t irq, unsigned int core)
+{
+	if (!desc) {
+		io_reg_update_bits(MTK_ADSP_IRQ_EN, BIT(irq), BIT(irq));
+	} else {
+		switch (desc->irq) {
+		case MTK_DSP_IRQ_MAILBOX:
+			io_reg_update_bits(MTK_ADSP_IRQ_EN, BIT(desc->irq), BIT(desc->irq));
+			break;
+		default:
+			tr_err(&int_tr, "Invalid interrupt %d", desc->irq);
+			return;
+		}
+	}
+}
+
+static uint32_t mtk_irq_group_pending_status(uint32_t irq)
+{
+	uint32_t irq_status = 0;
+
+	if (irq == MTK_DSP_IRQ_MAILBOX &&
+	    (io_reg_read(MTK_ADSP_IRQ_STATUS) & BIT(MTK_DSP_IRQ_MAILBOX))) {
+		irq_status = io_reg_read(MTK_MBOX_IRQ_IN) & MTK_DSP_MBOX_MASK;
+	}
+
+	return irq_status;
+}
+
+static uint32_t mtk_get_pending_index(uint32_t current, uint32_t *next)
+{
+	uint32_t index;
+
+	if (current == 0)
+		return PENDING_IRQ_INDEX_MAX;
+
+	/* ffs returns one plus the index of the least significant 1-bit of input int */
+	index = ffs(current) - 1;
+
+	/* remove the handling index from current pending status */
+	*next = current & ~(1ull << index);
+
+	return index;
+}
+
+static inline void mtk_handle_group_pending_irq(struct irq_cascade_desc *cascade,
+						uint32_t line_index, uint32_t status)
+{
+	int core = cpu_get_id();
+	struct list_item *clist;
+	struct irq_desc *child = NULL;
+	uint32_t idx;
+	uint32_t next_status;
+	bool handled;
+
+	idx = mtk_get_pending_index(status, &next_status);
+	while (idx < PENDING_IRQ_INDEX_MAX) {
+		handled = false;
+
+		spin_lock(&cascade->lock);
+		list_for_item(clist, &cascade->child[idx].list) {
+			child = container_of(clist, struct irq_desc, irq_list);
+
+			if (child->handler && (child->cpu_mask & 1 << core)) {
+				child->handler(child->handler_arg);
+				handled = true;
+			}
+		}
+		spin_unlock(&cascade->lock);
+
+		if (!handled) {
+			tr_err(&int_tr, "Not handle irq %u in group %u",
+			       idx, line_index);
+		}
+
+		idx = mtk_get_pending_index(next_status, &next_status);
+	}
+}
+
+static inline void mtk_irq_group_handler(void *data, uint32_t line_index)
```

Comment:
It's because some platforms also use it. I thought that maybe it's the way for IRQ handling at here.

for example,
[https://github.com/thesofproject/sof/blob/main/src/drivers/amd/renoir/interrupt.c#L128](https://github.com/thesofproject/sof/blob/main/src/drivers/amd/renoir/interrupt.c#L128)

[https://github.com/thesofproject/sof/blob/main/src/drivers/imx/interrupt-irqsteer.c#L324](https://github.com/thesofproject/sof/blob/main/src/drivers/imx/interrupt-irqsteer.c#L324)

Reply:
Interesting @lgirdwood this seems rather inflated, what are your thoughts here?

Reply:
Yeah, these could be optimized. Lets merge this first and get the build results as we could make the same change to all platforms. @tinghan-shen pls can you try this change and report any size changes to the binary. Thanks.

Reply:
I think NXP would see the biggest effect since they have multiple IRQs @dbaluta

Reply:
I tried to build mt8186 on both gcc11 and xcc, and compare the size of sof.ri.
No difference on gcc11.
64 bytes larger on xcc after remove "inline".